content/en/blog/_posts/2024-11-11-dynamic-kubernetes-api-server-for-cozystack/index.md

```diff
@@ -0,0 +1,454 @@
+
+### Disable etcd support
+
+In our case, we don't need it since all resources are stored directly in the Kubernetes API.
+
+You can disable etcd options by passing nil to RecommendedOptions.Etcd:
+
+- [Disabling etcd options](https://github.com/aenix-io/cozystack/blob/003edf8cf0a419bd67cd822d61ff806db49e7026/pkg/cmd/server/start.go#L70)
+
+### Generate a common resource type
+
+We called it `Application`, and it looks like this:
+
+- [Application type definition](https://github.com/aenix-io/cozystack/blob/003edf8cf0a419bd67cd822d61ff806db49e7026/pkg/apis/apps/v1alpha1/types.go)
+
+This is a generic type used for any application type, and its handling logic is the same for all charts.
+
+### Configure configuration loading
+
+Since we want to configure our extension api-server via a config file, we formed the config structure in Go:
+
+- [Config type definition](https://github.com/aenix-io/cozystack/blob/003edf8cf0a419bd67cd822d61ff806db49e7026/pkg/config/config.go)
+
+We also modified the resource registration logic so that the resources we create are registered in scheme with different Kind values:
+
+- [Dynamic Resource registration](https://github.com/aenix-io/cozystack/blob/003edf8cf0a419bd67cd822d61ff806db49e7026/pkg/apis/apps/v1alpha1/register.go#L63-L77)
+
+As a result, we got a config where you can pass all possible types and specify what they should map to:
+
+- [ConfigMap example](https://github.com/aenix-io/cozystack/blob/003edf8cf0a419bd67cd822d61ff806db49e7026/packages/system/cozystack-api/templates/configmap.yaml)
+
+### Implement our own registry
+
+To store state not in etcd but translate it directly into Kubernetes HelmRelease resources (and vice versa),
+we wrote conversion functions from Application to HelmRelease and from HelmRelease to Application:
+
+- [Conversion functions](https://github.com/aenix-io/cozystack/blob/003edf8cf0a419bd67cd822d61ff806db49e7026/pkg/registry/apps/application/rest.go#L920-L991)
+
+We implemented logic to filter resources by chart name, sourceRef, and prefix in the HelmRelease name:
+
+- [Filtering functions](https://github.com/aenix-io/cozystack/blob/003edf8cf0a419bd67cd822d61ff806db49e7026/pkg/registry/apps/application/rest.go#L747-L784)
+
+Then, using this logic, we implemented the methods `Get()`, `Delete()`, `List()`, `Create()`.
+
+You can see the full example here:
+
+- [Registry Implementation](https://github.com/aenix-io/cozystack/blob/003edf8cf0a419bd67cd822d61ff806db49e7026/pkg/registry/apps/application/rest.go)
+
+At the end of each method, we set the correct Kind and return an `unstructured.Unstructured{}` object
+so that Kubernetes serializes the object correctly. Otherwise,
+it would always serialize them with `kind: Application`, which we don't want.
+
+
+## What did we achieve?
+
+In Cozystack, all our types from the ConfigMap are now available in Kubernetes as-is:
+
+```shell
+# kubectl api-resources | grep cozystack
```

Comment:
[Separate commands from output](https://kubernetes.io/docs/contribute/style/style-guide/#separate-commands-from-output) (and don't include the prompt)